keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,730 @@
+	* Different information from different sources
+	* Confusion: Unclear origin of a given metric
+	* Performance degradations (increased CPU / Memory / etc) [xref][perf-issue]
+* Stats should be reported by the container runtime which knows behavior of the container/pod the best.
+* cAdvisor only supports runtimes that run processes on the host, not e.g. VM based runtime like Kata Containers.
+* cAdvisor only supports linux containers, not Windows ones.
+* cAdvisor has big list of external dependencies since it needs to support multiple container runtimes itself - leads to large support surface.
+
+[perf-issue]: https://github.com/kubernetes/kubernetes/issues/51798
+
+
+### Goals
+
+* Rely on a single component to provide the container and pod level metrics for the Summary API.
+* Improve performance and reduce confusion on metrics collection in the Kubelet.
+* Do not introduce breaking changes to the Summary API.
+* Eliminate dependencies on container runtime clients used by cAdvisor.
+* Move or deprecate the stats exclusively reported by `/metrics/cadivsor` to be collected (and maybe reported) by the CRI implementation.
+
+### Non-Goals
+
+- Have CRI support Volume Plugin stats (responsibility of the Kubelet).
+- Have CRI support Ephemeral Storage stats (responsibility of the Kubelet).
+- Have CRI support metrics of the rootfs (responsibility of cAdvisor).
+- Have CRI support Accelerator metrics stats (deprecated already in summary API).
+- Have CRI support UserDefinedMetrics.
+- Redefine which stats are reported in the Summary API (minus those that are being deprecated or not worth supporting in CRI)
+- Propose alternatives to the Summary API
+
+## Proposal
+
+This KEP encompasses three different components, all related to the goal of preventing duplicated work in gathering stats, as well as clarifying which component is responsible for what. These three pieces will be described in terms of the three components in play: CRI runtime, Kubelet, cAdvisor:
+
+The high level steps to achieve desired goal are:
+
+1. The [`ContainerStats`](https://github.com/kubernetes/cri-api/blob/release-1.20/pkg/apis/runtime/v1/api.proto#L1303-L1312) CRI message needs to be extended to include all of the fields from Summary's API's [`ContainerStats`][#summary-container-stats-object].
+2. Add CRI Pod Level Stats message to CRI protobuf that includes all [Pod Level Stats][#summary-pod-stats-object] metrics from Summary API.
+3. Add support for the new CRI additions in supported container runtimes (CRI-O and containerd).
+4. Switch Kubelet's CRI stats provider from querying container and pod level stats from cAdvisor to newly added CRI pod and container level stats
+5. cAdvisor should stop collecting container and pod level stats. If any other components need container or pod level stats from cAdvisor, the CRI implementation should be queried instead.
+
+This will be described in more detail in the [design details section][#design-details]
```

Comment:
These links are broken right now. (should use ()'s instead of []'s I think)

Reply:
ah you're right thanks